### PR TITLE
Project: Enumerable methods - Milestones

### PR DESCRIPTION
**Pull request by Andrik27001**
Ruby Project 2 **"Enumerable methods"**, focused on understanding enumerable methods in Ruby by rebuilding them was developed on the followed steps:

**Milestone 1 & 2:**
- Added tools (Linters, README, gitignore)
- Rebuild "my_each", "my_each_with_index" and my_select Methods.

**Milestone 3 & 4:**
- Rebuild "my_all?", "my_any?" and "my_none?" Methods
- Rebuild "my_count", "my_map" and my_inject Methods.

**Milestone 5 & 6:**
- Modified my_map method to take either a proc or a block
- Created multiply_els to test my_inject
- Fixed Linters.

### CODE REVIEW

**Review by Temmarie — CHANGES_REQUESTED** [review]
### Changes Required! :recycle:
Hi @bashforger  and @Andrik27001  ! 👋 👋 👋 

### You have done well so far :clap: However, there are some changes you have to make for your project to be approved. Kindly work on them and submit them for another code review.

**Recommended**

- **.my_each, .my_each_with_index, my_select, .map, should return an enumerator when no block is given.**

*  #### My_all?
* [x]  When a class is passed as an argument should return true if all of the collection is a member of such class. For example:
`[1, 'dog', []].my_all?(Integer)` - should return false

* [x] When a pattern other than Regex or a Class is given should return true if all of the collection matches the pattern. For example:
`['dog','door','dish'].my_all?(/d/)` - should return true

* [x] When a Regex is passed as an argument should return true if all of the collection matches the pattern. For example:
`[3, 3, 3].my_all?(3)` - should return true

*  #### My_any?
* [x] When no block or argument is given should return true if the block ever returns a value other than false or nil. For example:
`[1, 2, 3, 4].my_any?{|num| num > 4}` - should return false

* [x] When no block or argument is given should return true if at least one of the collection is not false or nil. For example:
`[nil, false, nil, false].my_any?` - should return false

* [x] When a class is passed as an argument should return true if at least one of the collection is a member of such class. For example:
`[1, 2, 3, 4, 5].my_any?(Integer)` - should return true

* [x]  When a Regex is passed as an argument should return false if none of the collection matches the Regex. For example:
`['dog','door','dish'].my_any?(/z/)` - should return false

* [x] When a pattern other than Regex or a Class is given should return false if none of the collection matches the pattern. For example:
`['dog', 'bird', 'fish'].my_any?('dog')` - should return true

*  #### My_none?
* [x] Should return true if the block never returns true for all elements. For example: 
`[1, 2, 3, 4].my_none?{|num| num > 4}` - should return true

* [x] When no block or argument is given should return true only if none of the collection members is true. For example:
`[nil, false, nil, false].my_none?` - should return true

* [x] When a class is passed as an argument should return true if none of the collection is a member of such class. For example:
[1, 2 ,3].my_none?(String) - should return false

* [x]  When a Regex is passed as an argument should return true only if none of the collection matches the Regex. For example:
`['dog','door','dish'].my_none?(/z/)` - should return true

* [x] When a pattern other than Regex or a Class is given should return true only if none of the collection matches the pattern. For example:
`['dog', 'bird', 'fish'].my_none?(5)` - should return true

*  #### My_inject
* [x] Should combine all elements of enum by applying a binary operation, specified by a block. For example: 
`(1..7).my_inject(4){|prod, n| prod*n}` - should return 20160
 
* [x] When a symbol is specified should combine each element of the collection by applying the symbol as a named method. For example: 
`[1, 2, 3, 4, 5].my_inject(:+)` - should return 15


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Review by Temmarie — CHANGES_REQUESTED** [review]
### Changes Required! :recycle:
Hi @bashforger and @Andrik27001 ! 👋 👋 👋 
### You have done well so far :clap: However, you did not finish all the previous review changes. Kindly work on them and submit them for another code review.


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Review by Temmarie — CHANGES_REQUESTED** [review]
### Changes Required! :recycle:
Hi @bashforger and @Andrik27001 ! 👋 👋 👋 

### You have done well so far :clap: However, you did not finish all the previous review changes. Kindly work on them and submit them for another code review.

**_You have made most of the necessary changes, the major issue with the other methods is "my_each", as you've used it in other methods, it causes errors there. Please make sure that `my_each` works just like `each` [here](https://ruby-doc.org/core-2.7.2/Enumerable.html)._**

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Comment by 0x1100010010** [comment]
We have verified `my_each` with the original `each` method. I am unable to understand what are you pointing to. Can you please pinpoint the line or section of code?

**Review by Temmarie — CHANGES_REQUESTED** [review]
### Changes Required! :recycle:
Hi @bashforger and @Andrik27001 ! 👋 👋 👋 

### You have done well so far :clap: However, you did not finish all the previous review changes. Refer to the changes requested [here](https://github.com/bashforger/enumerable-methods-ruby/pull/2#pullrequestreview-535773669) .Kindly work on them and submit them for another code review.

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] when a block is given when self is a range it should return the range itself after calling the given block once for each element in self

```Ruby
  range.my_each_with_index(&block) # returns undefined local variable or method `length' for 5..50:Range
```

- [x] when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

```Ruby
  enum.my_each_with_index(&block) # returns undefined local variable or method `length' for 5..50:Range
```

- [x] when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self

```Ruby
  block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
  my_each_with_index(&block)
  
  # expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
  # got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"

```

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] it should return false if the block returns false or nil::range

```Ruby
  range.my_all?(&false_block) # returns undefined local variable or method `length' for 5..50:Range
```

- [x] when no block or argument is given it should return false when one of the collection members are false or nil

```Ruby
  false_array.my_all?

  # expected false
  # got true
```

- [x] when a class is passed as an argument it should return true if all of the collection is a member of such class::Numeric

```Ruby
  numbers.my_all?(Numeric)

  # expected true
  # got false
```

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] when a class is passed as an argument it should return true if at least one of the collection is a member of such class::Numeric

```Ruby
  array.my_any?(Numeric)

  # expected true
  # got false
```

- [x] when a pattern other than Regex or a Class is given it should return false if none of the collection matches the pattern

```Ruby
   words.my_any?('cat')

  # expected false
  # got nil
```

- [x] when a pattern other than Regex or a Class is given it should return true if any of the collection matches the pattern

```Ruby
  words.my_any?('cat')

  # expected true
  # got nil
```

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] when a Regex is passed as an argument it should return true if none of the collection matches the Regex

```Ruby
  words.my_none?(/z/)

  # expected true
  # got nil
```

- [x] when a Regex is passed as an argument it should return false if any of the collection matches the Regex

```Ruby
  words.my_none?(/d/)

  # expected false
  # got nil
```

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] it should return the number of items in enum through enumeration::range

```Ruby
  range.my_count # undefined local variable or method `length' for 5..50:Range
```

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] raises a "LocalJumpError" when no block or argument is given


```Ruby

  array.my_inject 

  # expected LocalJumpError, got #<TypeError: nil is not a symbol nor a string> with backtrace:
           #  `block in my_inject'
           #  `my_each'
           #  `my_inject'

```

- [x] when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range

```Ruby
Failure/Error: actual = range.my_inject { |prod, n| prod * n }
      
      NoMethodError:
        undefined method `[]' for 5..50:Range
       `my_inject'
```

- [x] when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range


```Ruby
  range.my_inject(:*)

  # expected: 1267253883404724085150525340252698701849068398706688000000000000
  # got: 0
```

File: main.rb
Comment (SpaYco) [inline main.rb]: - [x] #multiply_els should be outside of the module

**Review by SpaYco — CHANGES_REQUESTED** [review]
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

variables used in testing: 

```Ruby
  ARRAY_SIZE = 100
  LOWEST_VALUE = 0
  HIGHEST_VALUE = 9

  array = Array.new(ARRAY_SIZE = rand(LOWEST_VALUE...HIGHEST_VALUE) } 
  block = proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } 
  words = %w[dog door rod blade] 
  range = Range.new(5, 50) 
  hash = { a: 1, b: 2, c: 3, d: 4, e: 5 } 
  numbers = [1, 2i, 3.14] 
  array_clone = array.clone 

```


## Required Changes ♻️ 

- [x] I addressed the required changes below 👇 

### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

**Review by Temmarie — CHANGES_REQUESTED** [review]
### Changes Required! :recycle:
Hi @bashforger and @Andrik27001 ! 👋 👋 👋 

### You have done well so far :clap: However, you did not finish all the previous review changes. Kindly work on them and submit them for another code review.

__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

File: main.rb
Comment (0x1100010010) [reply main.rb]: We have tried to fix it. I think the problem was there was no defined method for `Ranges`. It looks good now in our tests.

**Review by PhillipUg — APPROVED** [review]
# Status: Project Approved! 🥇 
Congratulations on making all the requested changes and getting this project approved. :+1:

You can merge this branch into the main branch when you're ready and go on to your next project.
:bulb:  Do not forget to submit your project completion form after merging this branch.

All the best team,

For any assistance, you can contact me on slack as **Phillip Musiime**

**Comment by 0x1100010010** [comment]
Thank you @PhillipUg

**Pull request merged**
